src/webgldrawer.js

```diff
@@ -0,0 +1,1120 @@
+ *
+ * - Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ *
+ * - Neither the name of CodePlex Foundation nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+(function( $ ){
+
+   /**
+    * @class WebGLDrawer
+    * @memberof OpenSeadragon
+    * @classdesc Default implementation of WebGLDrawer for an {@link OpenSeadragon.Viewer}.
+    * @param {Object} options - Options for this Drawer.
+    * @param {OpenSeadragon.Viewer} options.viewer - The Viewer that owns this Drawer.
+    * @param {OpenSeadragon.Viewport} options.viewport - Reference to Viewer viewport.
+    * @param {Element} options.element - Parent element.
+    * @param {Number} [options.debugGridColor] - See debugGridColor in {@link OpenSeadragon.Options} for details.
+    */
+
+    $.WebGLDrawer = class WebGLDrawer extends OpenSeadragon.DrawerBase{
+        constructor(options){
+           super(options);
+
+           this.destroyed = false;
+           // private members
+
+           this._TextureMap = new Map();
+           this._TileMap = new Map();
+
+           this._gl = null;
+           this._firstPass = null;
+           this._secondPass = null;
+           this._glFrameBuffer = null;
+           this._renderToTexture = null;
+           this._glFramebufferToCanvasTransform = null;
+           this._outputCanvas = null;
+           this._outputContext = null;
+           this._clippingCanvas = null;
+           this._clippingContext = null;
+           this._renderingCanvas = null;
+
+           // Add listeners for events that require modifying the scene or camera
+           this.viewer.addHandler("tile-ready", ev => this._tileReadyHandler(ev));
+           this.viewer.addHandler("image-unloaded", ev => this._imageUnloadedHandler(ev));
+
+           // this.viewer is set by parent constructor
+           // this.canvas is set by parent constructor, created and appended to the viewer container element
```

Comment:
The `canvas` HTML element is exposed as a public API, so that is created by `DrawerBase`. All the stuff internal to the webgl drawer implementation (setting up the actual webgl canvas, for example) is done in `_setupCanvases`.

Reply:
Oh, of course... That makes sense!